ecscollector/collector.go

```diff
@@ -0,0 +1,231 @@
+		"ecs_network_receive_dropped",
+		"Network packets dropped in recieving.",
+		networkLabels, nil)
+
+	networkRxErrorsDesc = prometheus.NewDesc(
+		"ecs_network_receive_errors",
+		"Network errors in recieving.",
+		networkLabels, nil)
+
+	networkTxBytesDesc = prometheus.NewDesc(
+		"ecs_network_transmit_bytes",
+		"Network transmitted in bytes.",
+		networkLabels, nil)
+
+	networkTxPacketsDesc = prometheus.NewDesc(
+		"ecs_network_transmit_packets",
+		"Network packets transmitted.",
+		networkLabels, nil)
+
+	networkTxDroppedDesc = prometheus.NewDesc(
+		"ecs_network_transmit_dropped",
+		"Network packets dropped in transmit.",
+		networkLabels, nil)
+
+	networkTxErrorsDesc = prometheus.NewDesc(
+		"ecs_network_transmit_errors",
+		"Network errors in transmit.",
+		networkLabels, nil)
+)
+
+var labels = []string{
+	"container",
+}
+
+var cpuLabels = append(
+	labels,
+	"mode", // user, kernel, system
+)
+
+var networkLabels = append(
+	labels,
+	"device",
+)
+
+// NewCollector returns a new Collector that queries ECS metadata server
+// for ECS task and container metrics.
+func NewCollector(client *ecsmetadata.Client) prometheus.Collector {
+	return &collector{client: client}
+}
+
+type collector struct {
+	client *ecsmetadata.Client
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- numProcsDesc
+	ch <- cpuTotalDesc
+	ch <- cpuSystemDesc
+	ch <- cpuOnlineDesc
+	ch <- memUsageDesc
+	ch <- memMaxUsageDesc
+	ch <- memLimitDesc
+	ch <- networkRxBytesDesc
+	ch <- networkRxPacketsDesc
+	ch <- networkRxDroppedDesc
+	ch <- networkRxErrorsDesc
+	ch <- networkTxBytesDesc
+	ch <- networkTxPacketsDesc
+	ch <- networkTxDroppedDesc
+	ch <- networkTxErrorsDesc
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	ctx := context.Background()
+	metadata, err := c.client.RetrieveTaskMetadata(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve metadata: %v", err)
+		return
+	}
+	stats, err := c.client.RetrieveTaskStats(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve container stats: %v", err)
+		return
+	}
+	for _, container := range metadata.Containers {
+		s := stats[container.DockerID]
+		if s == nil {
+			log.Printf("Couldn't find container with ID %q in stats", container.DockerID)
+			continue
+		}
+
+		labelVals := []string{
+			container.Name,
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			cpuTotalDesc,
+			prometheus.CounterValue,
+			s.CPUStats.CPUUsage.TotalUsage,
+			append(labelVals, "total")...,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			cpuTotalDesc,
+			prometheus.CounterValue,
+			s.CPUStats.CPUUsage.UsageInKernelMode,
+			append(labelVals, "kernel")...,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			cpuTotalDesc,
+			prometheus.CounterValue,
+			s.CPUStats.CPUUsage.UsageInUserMode,
+			append(labelVals, "user")...,
+		)
+		ch <- prometheus.MustNewConstMetric(
+			cpuSystemDesc,
+			prometheus.CounterValue,
+			s.CPUStats.SystemCPUUsage,
+			labelVals...,
+		)
```

Comment:
I'm not sure what this is for, it seems to overlap with the above "kernel" mode.

In the example data I found, it has a completely different scale, so I have no idea what the unit is here. Maybe it's represented in milliseconds or microseconds?

The documentation on this API is very lacking.